apps/emqx_durable_storage/src/emqx_ds.erl

```diff
@@ -94,9 +94,11 @@
 -type next_result() :: next_result(iterator()).
 
 %% Timestamp
+%% Each message must have unique timestamp.
```

Comment:
Note: storage lts introduces a "unique integer" https://github.com/emqx/emqx/blob/5bc67cb288f6953e73654ce1784b22d98b2639b4/apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl#L446
to keep the original timestamp.

I think creation of this number can be moved into the replication layer, and made deterministic (smth. like raft batch seqno + message index in the batch). This way we'll avoid storing two unique IDs.

Reply:
> this PR shifts the responsibility of making the timestamps unique to the application

To the replication layer actually.

> We shouldn't move the responsibility of making message timestamps unique to the API consumer

Agree. This PR doesn't move the responsibility though. The replication layer modifies message timestamps to ensure uniqueness. API consumer will definitely be surprised though, but it's a PoC-grade hack at this point.

> we can order the messages according to the leader's clock

They are already ordered according to the leader's clock (to be more precise, shard's consensus group tracks strictly monotonically increasing clock).